File: KEYBOARD_ORBY_V3/RASPBERRY_PROJECT/ORBY_V4/OrbyGUI/electron/log.js
// Registro en disco de lo que le pasa a la conexión con el teclado.
//
// En desarrollo basta con el terminal, pero la app instalada no tiene consola
// que mirar: cuando alguien dice "no me detecta el teclado", no había ni una
// línea con la que saber si el puerto no se abrió, si el teclado no se presentó
// o si la conexión se murió sin avisar. Esto deja un rastro que se puede pedir.
//
// Solo el ciclo de vida de la conexión: ni la telemetría ni los comandos, que
// serían megas al día y aquí no aportan.

const fs = require('fs');
const path = require('path');

// Fuera de Electron (las pruebas sueltas de tools/ cargan serial.js en Node a
// pelo) no hay carpeta de datos de usuario: entonces esto es solo la consola.
let app = null;
try { ({ app } = require('electron')); } catch { /* Node pelado */ }

// Se parte en dos ficheros de este tamaño: el actual y el anterior. Con eso
// sobra para ver qué pasó en el último arranque sin dejar el disco lleno.
const MAX_BYTES = 256 * 1024;

let file = null;

function target() {
  if (!file && app) file = path.join(app.getPath('userData'), 'orby.log');
  return file;
}

function rotate() {
  try {
    if (fs.statSync(target()).size < MAX_BYTES) return;
    try { fs.unlinkSync(`${target()}.1`); } catch { /* no había anterior */ }
    fs.renameSync(target(), `${target()}.1`);
  } catch { /* no existe todavía, o no se puede rotar: se sigue escribiendo */ }
}

function write(line) {
  if (!target()) return;
  const stamp = new Date().toISOString();
  try {
    rotate();
    fs.appendFileSync(target(), `${stamp} ${line}\n`, 'utf8');
  } catch { /* si no se puede escribir el registro, no se rompe nada por ello */ }
}

// Sale también por la consola: en desarrollo es donde se mira.
function log(...parts) {
  const line = parts.join(' ');
  console.log(line);
  write(line);
}

module.exports = { log, file: target };

File: KEYBOARD_ORBY_V3/RASPBERRY_PROJECT/ORBY_V4/OrbyGUI/src-tauri/src/log.rs
// Registro en disco de lo que le pasa a la conexión con el teclado.
//
// Es el equivalente de `electron/log.js`, y escribe en el **mismo** fichero:
// `%APPDATA%\OrbyGUI\orby.log`. En desarrollo basta con el terminal, pero la app
// instalada no tiene consola —`windows_subsystem = "windows"` en `main.rs`—, así que sin
// esto un `eprintln!` no se puede leer en ningún sitio. Cuando alguien dice «se me
// desconecta solo», este fichero es la única forma de saber si lo soltó una escritura,
// una lectura o el vigilante.
//
// Solo el ciclo de vida de la conexión: ni la telemetría ni los comandos, que serían
// megas al día y aquí no aportan.

use std::io::Write;
use std::path::PathBuf;
use std::sync::{Mutex, OnceLock};

/// Se parte en dos ficheros de este tamaño: el actual y el anterior. Con eso sobra para
/// ver qué pasó en el último arranque sin dejar el disco lleno.
const MAX_BYTES: u64 = 256 * 1024;

/// El hilo del puerto no es el único que registra (el del firmware también), y dos
/// `append` a la vez pueden entrelazar media línea con media línea.
fn cerrojo() -> &'static Mutex<()> {
    static CERROJO: OnceLock<Mutex<()>> = OnceLock::new();
    CERROJO.get_or_init(|| Mutex::new(()))
}

/// Junto a `orby-config.json`, y por el mismo motivo que aquel no sale de
/// `app_config_dir()`: el identificador de la app es `com.orby.gui` y dejaría el rastro
/// en `%APPDATA%\com.orby.gui`, que no es donde nadie lo va a buscar.
fn ruta() -> Option<PathBuf> {
    let appdata = std::env::var_os("APPDATA")?;
    Some(PathBuf::from(appdata).join("OrbyGUI").join("orby.log"))
}

fn rotar(p: &PathBuf) {
    let Ok(meta) = std::fs::metadata(p) else { return };
    if meta.len() < MAX_BYTES {
        return;
    }
    let anterior = p.with_extension("log.1");
    let _ = std::fs::remove_file(&anterior);
    let _ = std::fs::rename(p, &anterior);
}

/// Escribe una línea con su marca de tiempo. **Nunca falla hacia fuera**: si no se puede
/// registrar, eso no puede romper la conexión que se estaba registrando.
pub fn escribir(linea: &str) {
    // En desarrollo se mira el terminal, así que sale por los dos sitios.
    eprintln!("{linea}");

    let Some(p) = ruta() else { return };
    let _guarda = cerrojo().lock().unwrap_or_else(|e| e.into_inner());

    if let Some(dir) = p.parent() {
        let _ = std::fs::create_dir_all(dir);
    }
    rotar(&p);

    let marca = chrono::Local::now().format("%Y-%m-%dT%H:%M:%S%.3f");
    if let Ok(mut f) = std::fs::OpenOptions::new().create(true).append(true).open(&p) {
        let _ = writeln!(f, "{marca} {linea}");
    }
}

/// `log!("[serie] ...")`, para no tener que montar el `format!` en cada llamada.
macro_rules! log {
    ($($arg:tt)*) => { crate::log::escribir(&format!($($arg)*)) };
}

pub(crate) use log;
